### PR TITLE
Avoid storing empty DetSet containers in the pixel digi collection

### PR DESCRIPTION
**Pull request by ferencek**
#### PR description:

This PR addresses the issue with excessive LogWarning printouts in the pixel clusterizer reported in https://github.com/cms-sw/cmssw/pull/36326#issuecomment-984417751. This PR is an alternative to https://github.com/cms-sw/cmssw/pull/36968 and tries to fix the problem at the source, in the RawToDigi code. This is a technical fix so no impact on physics is expected.

#### PR validation:

The code compiles and was tested to successfully run with the MinimumBias workflow 139.001.

@tsusa @mmusich

### CODE REVIEW

**Comment by mmusich** [comment]
please test

**Comment by mmusich** [comment]
type bug-fix

**Comment by VinInn** [comment]
There was a rational behind the current behaviour:  an empty detset signals that it has been decoded and it is empty. no detset signals that was not read...  (or absent in the redout?)

**Comment by jpata** [comment]
DQM detects some differences in 136.731 / Pixel / Barrel:
image

are these expected?

**Comment by ferencek** [comment]
> There was a rational behind the current behaviour: an empty detset signals that it has been decoded and it is empty. no detset signals that was not read... (or absent in the redout?)

Hi @VinInn, ok, I see, so the empty DetSets are a sort of diagnostic of corrupt data coming from certain modules. But isn't the purpose of the error collection `edm::DetSetVector<SiPixelRawDataError>` precisely to store this type of information. If I look at this comment https://github.com/cms-sw/cmssw/blob/CMSSW_12_2_0/EventFilter/SiPixelRawToDigi/plugins/SiPixelRawToDigi.cc#L241 right before `PixelDataFormatter::interpretRawData(...)` is called, I would naively guess there would be no empty DetSets stored in the digi collection.

**Comment by ferencek** [comment]
> DQM detects some differences in 136.731 / Pixel / Barrel: image
> 
> are these expected?

Hmm, I would say these are not expected based on the code changes. I also see some small differences in 4.53 / Pixel / Endcap. I will need to check these more closely (unless we decide to drop the PR altogether).

**Comment by mmusich** [comment]
> Hmm, I would say these are not expected based on the code changes. I also see some small differences in 4.53 / Pixel / Endcap. I will need to check these more closely (unless we decide to drop the PR altogether).

somehow interesting these show only in the phase-0 wfs.

**Comment by ferencek** [comment]
I had a closer look at wf 136.731 where I see an empty DetSet appearing
```
Begin processing the 91st record. Run 274199, Event 40350490, LumiSection 21 on stream 0 at 23-Feb-2022 23:27:58.280 CET
%MSG-w clusterizeDetUnit():   SiPixelClusterProducer:siPixelClustersPreSplitting@cpu  23-Feb-2022 23:27:58 CET Run: 274199 Event: 40350490
No digis to clusterize
%MSG
```
and similarly in wf 4.53
```
Begin processing the 15th record. Run 194533, Event 462529546, LumiSection 329 on stream 0 at 24-Feb-2022 08:25:21.711 CET
%MSG-w clusterizeDetUnit():   SiPixelClusterProducer:siPixelClustersPreSplitting@cpu  24-Feb-2022 08:25:21 CET Run: 194533 Event: 462529546
No digis to clusterize
%MSG
Begin processing the 47th record. Run 194533, Event 462781441, LumiSection 329 on stream 0 at 24-Feb-2022 08:26:35.774 CET
%MSG-w clusterizeDetUnit():   SiPixelClusterProducer:siPixelClustersPreSplitting@cpu  24-Feb-2022 08:26:35 CET Run: 194533 Event: 462781441
No digis to clusterize
%MSG
Begin processing the 93rd record. Run 194533, Event 462831597, LumiSection 329 on stream 0 at 24-Feb-2022 08:28:28.169 CET
%MSG-w clusterizeDetUnit():   SiPixelClusterProducer:siPixelClustersPreSplitting@cpu  24-Feb-2022 08:28:28 CET Run: 194533 Event: 462831597
No digis to clusterize
%MSG
```
Specifically for wf 136.731, I checked that the total number of digis in each event remains unchanged before and after the fix and the difference in the SUMOFF_digis_Barrel appearing in ladder 65
SUMOFF_ndigis_Barrel_136 731
disappears once Run: 274199 Event: 40350490 is skipped
SUMOFF_ndigis_Barrel_136 731_skipBadEvent

So the differences appearing in the SUMOFF_digis_Barrel are an artifact of how the number of digis in a given module is counted. There are three options:
* DetId appearing with N pixel hits: N digis in the corresponding DetSet
* DetId appearing with no valid hits: Empty DetSet, 0 digis counted for that DetId
* DetId missing: Nothing done for that DetId

In other words, the differences we see is equivalent to comparing the average of (0, 5, 7) and the average of (5, 7) which will, of course, be different.

The multiplicity of higher level objects such as clusters and RecHits remains unchanged
SUMOFF_nRecHits_Barrel_136 731

So to conclude, this fix is indeed transparent to physics.

P.S. @mmusich made an interesting observation that differences seem to be showing up only for Phase 0 wfs. This is indeed confusing because empty DetSets were earlier observed by Marco in the Phase 1 wf 139.001. However, there are differences in the content of the DQM files for Phase 0
DQM_Phase0
and Phase 1
DQM_Phase1
so the ndigis plot showing differences in the Phase 0 wfs is actually not available for Phase 1.

File: EventFilter/SiPixelRawToDigi/src/PixelDataFormatter.cc [hunk 1]
```diff
@@ -204,6 +202,14 @@ void PixelDataFormatter::interpretRawData(
...
+      nrawId = rawId;
+      detDigis = &digis.find_or_insert(rawId);
+      if ((*detDigis).empty()) {
+        (*detDigis).data.reserve(32);  // avoid the first relocations
+      }
+    }
+
     GlobalPixel global = rocp->toGlobal(*local);  // global pixel coordinate (in module)
     (*detDigis).data.emplace_back(global.row, global.col, adc);
     LogTrace("") << (*detDigis).data.back();
```
Comment (perrotta) [inline EventFilter/SiPixelRawToDigi/src/PixelDataFormatter.cc:215]: Would a protection be needed here (see static analyzer report)?
```suggestion
    if (detDigis) {
      GlobalPixel global = rocp->toGlobal(*local);  // global pixel coordinate (in module)
      (*detDigis).data.emplace_back(global.row, global.col, adc);
      LogTrace("") << (*detDigis).data.back();
    }
```

Comment (ferencek) [reply EventFilter/SiPixelRawToDigi/src/PixelDataFormatter.cc:215]: Thanks, @perrotta, for the suggestion. Indeed, this is good to add. Perhaps it would also be good to add a LogError as an else follow-up because a null pointer at this point is normally not expected.

**Comment by tvami** [comment]
@cmsbuild , please test

**Comment by ferencek** [comment]
I applied Andrea's commit suggestion and amended it with a LogError printout.

**Comment by jpata** [comment]
@cmsbuild please test

**Comment by clacaputo** [comment]
+reconstruction

- PR addresses excessive LogWarning printouts in the pixel clusterizer reported in #36326 (comment)
- no reco changes expected or observed
- DMQ differences in `SUMOFF_ndigis_Barrel` explained in https://github.com/cms-sw/cmssw/pull/37035#issuecomment-1049601999
  - The multiplicity of higher level objects such as clusters and RecHits remains unchanged

**Comment by mmusich** [comment]
> and the conclusion was to merge both PRs with a small change in https://github.com/cms-sw/cmssw/pull/36968 where LogWarning is turned into LogError instead of LogDebug.

https://github.com/cms-sw/cmssw/pull/36968 has been changed accordingly.

**Comment by perrotta** [comment]
+1
